### PR TITLE
feat(oidc): draft implementation of OIDC server & client

### PR DESCRIPTION
**Pull request by rudyardrichter**
TODOs etc. moved into issue description.

### CODE REVIEW

File: bin/fence-create [hunk 1]
```diff
@@ -35,51 +33,18 @@ def delete_client_action(DB, client):
         print e.message
 
 
-def sync_dbgap(projects):
```
Comment (philloooo) [inline bin/fence-create]: why is this removed?

File: dev-requirements.txt [hunk 2]
```diff
@@ -2,4 +2,4 @@ pytest-cov==2.5.1
 codacy-coverage
 mock==2.0.0
 moto==1.1.24
--e git+https://github.com/uc-cdis/cdisutils-test.git@0.1.0#egg=cdisutilstest
```
Comment (philloooo) [inline dev-requirements.txt]: why are we decreasing the version?

File: fence/__init__.py [hunk 3]
```diff
@@ -42,10 +43,6 @@ def app_config(app, settings='fence.settings', root_dir=None):
     if root_dir is None:
         root_dir = os.path.dirname(
                 os.path.dirname(os.path.realpath(__file__)))
-    if 'AWS_CREDENTIALS' in app.config and len(app.config['AWS_CREDENTIALS']) > 0:
```
Comment (philloooo) [inline fence/__init__.py]: why is this removed?

File: bin/fence-create (hunk 1 above)
Comment (rudyardrichter) [reply bin/fence-create]: Should not be. Will fix.

File: fence/local_settings.example.py [hunk 4]
```diff
@@ -40,24 +38,14 @@
     }
 }
 
-# Configuration necessary for cirrus (Cloud Management Library)
-# https://github.com/uc-cdis/cirrus
-# will eventually be passed as params but cirrus looks for env vars right now
-os.environ["GOOGLE_API_KEY"] = ""
```
[the remark this answers is not in the document]
Comment (rudyardrichter) [reply fence/local_settings.example.py]: Yep, this (and some other stuff, see Phillis' comments) got scrapped accidentally as the result of the halfway-merge commit being reverted. Will fix.

File: fence/auth.py [hunk 5]
```diff
@@ -97,30 +96,46 @@ def wrapper(*args, **kwargs):
...
 def has_oauth(scope=None):
     scope = scope or set()
-    scope.update({'access'})
+    scope.update({'openid'})
     try:
-        user_api = capp.config['USER_API'] if capp.config.has_key('USER_API') else capp.config['HOSTNAME']
-        access_token = auth.validate_request_jwt(
-            aud=scope, user_api=user_api
-        )
+        claims = validate_jwt(aud=scope, purpose='id')
```
[the remark this answers is not in the document]
Comment (Avantol13) [reply fence/auth.py]: Should be `access`

**Comment by rudyardrichter** [comment]
@Avantol13 responded to your comments in 45625c7

File: fence/oidc/jwt_generator.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+class JWTGenerator(BearerToken):
+    """
+    Implement bearer token generation behavior as required for authlib.
+
+    Instances of ``JWTGenerator`` are callables which return the token
+    response (see ``__call__``).
+    """
+
+    ACCESS_TOKEN_EXPIRES_IN = 1200
```
Comment (Avantol13) [inline fence/oidc/jwt_generator.py]: I think these values are in local_settings? or settings? in any case, they should definitely be configurable outside of this class

File: fence/resources/storage/cdis_hmac.py [hunk 7]
```diff
@@ -17,10 +18,11 @@ def create_keypair(user, current_session, encryption_key, expire=86400):
     # default to 1 day, max to 30 days
     try:
         expire = int(expire)
-    except:
+    except ValueError:
```
Comment (Avantol13) [inline fence/resources/storage/cdis_hmac.py]: nice catch, I read an article somewhere that empty `except:` will even catch keyboard interrupts and `SystemExit` which is interesting and definitely not intended in most cases. anyway, fun fact of the day

File: fence/resources/storage/cdis_jwt.py [hunk 8]
```diff
@@ -1,25 +1,63 @@
...
 
 
 def create_session_token(
         keypair, expires_in, username=None, session_started=None,
         provider=None, redirect=None):
-    return token.generate_signed_session_token(keypair.kid, keypair.private_key,
-                                               expires_in, username,
-                                               session_started, provider, redirect)
+    return token.generate_signed_session_token(
+        keypair.kid, keypair.private_key, expires_in, username,
```
Comment (Avantol13) [inline fence/resources/storage/cdis_jwt.py]: heh, more of my old coding style, apologies, I think I'm avoiding it effectively now

File: fence/resources/storage/cdis_jwt.py [hunk 9]
```diff
@@ -1,25 +1,63 @@
...
+from fence.jwt.errors import JWTError
+from fence.jwt.validate import validate_jwt
+from fence.models import UserRefreshToken
+
+
+def create_id_token(
+        user, keypair, expires_in, client_id, audiences=None,
+        auth_time=None, max_age=None, nonce=None):
+    try:
+        return token.generate_signed_id_token(
```
Comment (Avantol13) [inline fence/resources/storage/cdis_jwt.py]: I overheard you guys talking about how these `token.generate` functions rely on flask context. Not something that needs to be done now but we could easily grab the context we need in these `create_x` functions and pass it down to make the generate functions decoupled from the context. anyway, just a thought

File: fence/resources/user/user_session.py [hunk 10]
```diff
@@ -230,7 +238,7 @@ def _create_access_token_cookie(app, response, user):
     access_token = generate_signed_access_token(
         keypair.kid, keypair.private_key, user,
         app.config.get('ACCESS_TOKEN_LIFETIME').seconds, scopes,
-        client_id=user.id, forced_exp_time=expiration
+        forced_exp_time=expiration
```
Comment (Avantol13) [inline fence/resources/user/user_session.py]: the client_id being the user was intentional here at one point, might not be an issue in the session anymore but now that I'm thinking about it: I think for API keys, the client_id should be the user.id, I'll have to look and see if that's the case (this is for some google group functionality in knowing who is acting on behalf of the user).

Comment (Avantol13) [reply fence/resources/user/user_session.py]: Edit: Api Key looks good, endpoint checks for client_id in request and grabs user.id and passes it into `create` as the client_id if it doesn't exist (e.g. we're talking to the user directly)

File: fence/oidc/jwt_generator.py (hunk 6 above)
Comment (rudyardrichter) [reply fence/oidc/jwt_generator.py]: Changed in 4a7edb9
